pbrd/pbr_vty.c

```diff
@@ -171,6 +171,28 @@ DEFPY(pbr_map_match_dst, pbr_map_match_dst_cmd,
 	return CMD_SUCCESS;
 }
 
+DEFPY(pbr_map_match_fwmark, pbr_map_match_fwmark_cmd,
+	"[no] match fwmark (1-4294967295)$fwmark",
```

Comment:
While I find this new command very useful, I don't think we should expose the concept of fwmark at the CLI level since this is a Linux-specific thing. What we could possibly do is to enclose this command under an `#ifdef linux` block, but I'm not sure if I like that approach. The ideal solution in my opinion would be to write generic commands for whatever PBR policies you need (e.g. match on src/dest port) and use fwmark under the hood when running FRR on Linux. This way we can have a consistent set of commands across all platforms supported by FRR (pbrd supports Linux only as of now, but this might change in the future).

@donaldsharp any thoughts on this?

Reply:
I guess I'm a bit confused a bit :) . What about this implementation isn't linux specific.  I can't speak to the abilities of other os's because I am ignorant here.  Is there even a generic ability to do this on other platforms?

Reply:
> What about this implementation isn't linux specific

The CLI commands, but maybe `pbr table range` can be considered as Linux-specific.

> Is there even a generic ability to do this on other platforms?

Probably yes, but each platform should have a specific API for this. So maybe we can't escape from having platform-specific commands on pbrd :(

Reply:
The problem with rule engine on Linux is that it cannot directly match on other criteria than src ip, dst ip and fwmark. So to achieve more advanced pbr scenarios, packets needs to be marked by netfilter. To make it more generic there should be a code to interface with netfilter to translate high level rules to netfilter marking entries. However netfilter is usually used for many other kind of tasks and usually there are rules in place created by user directly using iptables or other users-pace apps, It would be really painful from development and user perspective to keep all of it in sync, ensure proper order and avoid conflicts between rules from different sources. As long as frr doesn't have capability to fully manage filtering engines on a different platform it is hard to implement this without any tradeoffs. 

The proposed change to use fwmark directly is maybe a bit simplistic and platform specific approach, but it's sufficient for majority of users who just want to move management of existing policy based rules to pbrd (and for example make use of recursive route lookups which pbr provides) without loosing flexibility of already used methods and without forcing pbrd to interact with system components  it shouldn't interact with (netfilter, pf etc.).

I could add #ifdef macro to make sure fwmark command gets only compiled on Linux.

BTW. I did some research and it seems that ipproto and src/dst ports matches has been added not so long ago to iproute2 and linux kernel (released as part of 4.17), however these are not included within major distributions yet, and even if it was fwmark based pbr still has valid use cases. I will try out to extend pbrd for these new matches as well, but maybe as separate PR as I'm not sure how it will behave depending on kernel version.


https://github.com/torvalds/linux/commit/bfff4862653bb96001ab57c1edd6d03f48e5f035#diff-d51e5e718cf62762bb953f6198c33bca

https://github.com/shemminger/iproute2/commit/b2e8bf158460568ec5b48cba69f657f95891c901#diff-f44706923fd0a2a572f2207443b51ee5

Reply:
I'm guessing we'll need platform specific commands in pbrd -- given this, should we split the code into per platform modules in some way, so we don't have a mix of code for different platforms scattered around in files to be cleaned up later?

Reply:
I think pbr daemon should benefit from what the underlying system offers in terms of policy routing. If I take example of bgp flowspec, I reused fwmark option through zapi interface.

Reply:
If we restrict pbr to the subset of commands supported by all platforms we would be limiting it severely. I was talking to @donaldsharp about expanding  pbr to support fwmark/ports/etc. just a couple weeks ago. coincidentally, @n3wtype opened  #4460 and followed up with this PR. So, I agree with the general sentiment that it is OK to have   platform-specific commands, we just need to do it cleanly

Reply:
Instead if making the command a Linux only thing, I suggest we use a slightly more generic name, so that commands would be

[no] match mark (1-4294967295)$mark 

this would a standard pbr command that is always available. If the platform does support packet marking that result would be a working like:

"pbr marks are not supported on this platform"

Reply:
This suggestion works for me.

Reply:
noted

Reply:
@n3wtype thanks for the write-up, I think you made some very good points. I'll approve this PR now.

I also support Jafar's suggestion of using "mark" instead of "fwmark".

Reply:
I like this suggestion, as well.